User signup, login and profile view created

BarbDMC opened the pull request:
The changes applied were:

- The project was started with Vite.
- Tailwind dependency was added.
- The hooks useLogin and, useSignUp were created.
- The pages Login, SignUp, and Profile were created.
- Login, SignUp, and Profile TSX files were tested.
- Login and Signup responses were updated to return a user.

BarbDMC committed:
Versioning improved

BarbDMC committed:
React librery added to project

BarbDMC committed:
Folder structure changed

BarbDMC committed:
Feat: Home and Login pages created

BarbDMC committed:
feat: Enable login functionality

BarbDMC committed:
Fix: Login styles moved to its own file

BarbDMC committed:
Feat: Sign up view

BarbDMC committed:
Fix: signup styles moved to it's own file

BarbDMC committed:
Feat: Creating useSignup

BarbDMC committed:
Fix: error messages in login and signup

BarbDMC committed:
Feat: Creating user profile view

BarbDMC committed:
Feat: Redirect to profile view when login

BarbDMC committed:
Feat: Redirect to profile view after signup

BarbDMC committed:
Feat: Link to login page added to signup

BarbDMC committed:
Feat: Link to signup page added to login

BarbDMC committed:
refactor: rename components folder to pages

BarbDMC committed:
refactor: move profile styles to it's own file

BarbDMC committed:
refactor: migrate to vite

BarbDMC committed:
feat: add dark mode to home

BarbDMC committed:
test: add ts-jest types

BarbDMC committed:
test: fixing app tsx spec

BarbDMC committed:
test: adding sign up tests

BarbDMC committed:
Specs: app and profile missing specs added

dalejo96 commented on client/index.html:
```diff
@@ -0,0 +1,21 @@
+    />
+    <link rel="apple-touch-icon" href="/logo192.png" />
+    <link rel="manifest" href="/manifest.json" />
+    <title>React App</title>
```
Update this information

dalejo96 commented on client/package.json:
```diff
@@ -0,0 +1,56 @@
+    "postcss": "^8.4.33",
+    "tailwindcss": "^3.4.1",
+    "ts-jest": "^29.1.1"
+  },
```
Some `dependencies` and `devDependencies` are mixed, please check this.
It is important to have well-defined libraries because this increments the bundle size.
Let me know if you have any doubt or question.

dalejo96 commented on client/src/index.tsx:
```diff
@@ -0,0 +1,20 @@
+      <React.StrictMode>
+        <App />
+      </React.StrictMode>
+    </BrowserRouter>,
```
Let's make the src index the simple as possible, with just the `StricMode` and the `App` component, place the BrowserRouter inside `App`

dalejo96 commented on client/src/hooks/useLogin.ts:
```diff
@@ -0,0 +1,39 @@
+import { useState } from 'react';
+
+export const useLogin = () => {
+  const [loading, setLoading] = useState<boolean>(false);
```
It is kind of common to name boolean states as:
- isLoading
- canDownload
- canUpdate
- isValid

This improves code readability and also you do not have to specify the `<boolean>` generic, this can be inferred by TS.

dalejo96 commented on client/src/hooks/useLogin.ts:
```diff
@@ -0,0 +1,39 @@
+
+        setError(errorMessage);
+        setLoading(false);
+      }
```
What if the error is not an axiosErros?
You could do something like:
```ts
if (axios.isAxiosError(err)) {
   // some stuff with an axios error
}
// then some generic error handling
```
You can use `finally` statements in try/catch blocks or promises to trigger final actions for example:
```ts
try {
    // doing async things
    } catch (e) {
      // catching errors
    } finally {
      setIsLoading(false);
    }
```
This way you can avoid repeating code.

dalejo96 commented on client/src/hooks/useSignUp.ts:
```diff
@@ -0,0 +1,38 @@
+      }
+      
+    }
+  }
```
Same as above

dalejo96 commented on client/src/hooks/useSignUp.ts:
```diff
@@ -0,0 +1,38 @@
+
+import { useState } from 'react';
+import Axios from 'axios';
```
Let's standardize the axios usage:
```ts
import Axios from 'axios';
or
import axios from 'axios';
```
But do this for all apps

dalejo96 commented on client/src/interfaces/userInterfaces.ts:
```diff
@@ -0,0 +1,13 @@
+  surname:  string,
+  secondSurname?:  string,
+  birthday: string
+  email: string,
```
optional properties go at the bottom of the interface, there is an ESLint rule for validating this

dalejo96 commented on client/src/interfaces/userInterfaces.ts:
```diff
@@ -0,0 +1,13 @@
+
+export interface userDataInterface {
```
interfaces use capital letters at the beginning.

dalejo96 commented on client/src/pages/App/App.tsx:
```diff
@@ -0,0 +1,21 @@
+          <Route path="/" element={<Home />} />
+          <Route path="/signup" element={<SignUp />}/>
+          <Route path="/login" element={<Login />} />
+          <Route path="/profile" element={<Profile />} />
```
What if I go to `/aasdasdasd` or  `/account` ?
You have to create a 404 route with a wildcard route for example:
```ts
<Route path="*" element={<NotFound />} />
```

dalejo96 commented on client/src/pages/App/App.tsx:
```diff
@@ -0,0 +1,21 @@
+
+function App() {
+  return (
+    <div className="App">
```
Is this `div` needed, remember a div is an extra node that react has to rerender.

dalejo96 commented on client/src/pages/Home/Home.tsx:
```diff
@@ -0,0 +1,41 @@
+
+const Home = () => {
+  return (
+    <section className="home-container-background">
```
I guess the all purpose of having a tailwindcss is for inline styles and avoid extra files like `styles.css` what do you think @CristopherDavid ?

CristopherDavid replied:
I agree, but in this case maybe having them in one file is a more organized way to keep the styles. Tho I'm not a fan of tailwind (maybe you could use things like styled-components or a UI library)

dalejo96 commented on client/src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,59 @@
+
+export const Login = () => {
+  const navigate = useNavigate();
+  const [email, setEmail] = useState<string>("");
+  const [password, setPassword] = useState<string>("");
```
You can use a hook called `useReducer()` for handling this or even better install a forms management library such as react-hook-forms

CristopherDavid replied:
react-use-forms sounds good here, it'll help you keep this clean

dalejo96 commented on client/src/pages/Profile/Profile.tsx:
```diff
@@ -0,0 +1,35 @@
+import { useLocation } from "react-router-dom";
+
+export const Profile = () => {
+  const { state } = useLocation();
+  const { name, middleName, surname, secondSurname, email, birthday } = state;
```
This approach is not persistent, so this is ok for now, you have to improve in a future PR

dalejo96 commented on client/src/pages/SignUp/SignUp.tsx:
```diff
@@ -0,0 +1,128 @@
+import { useNavigate, Link } from 'react-router-dom';
+
+export const SignUp: React.FC = () => {
+  const [name, setName] = useState<string>("");
+  const [middleName, setMiddleName] = useState<string>("");
+  const [surname, setSurname] = useState<string>("");
+  const [secondSurname, setSecondSurname] = useState<string>("");
+  const [birthday, setBirthday] = useState<string>("");
+  const [email, setEmail] = useState<string>("");
+  const [password, setPassword] = useState<string>("");
```
use a reducer here

CristopherDavid replied:
Also, it will help you with validation and showing errors for each field

dalejo96 commented on client/specs/App.test.tsx:
```diff
@@ -0,0 +1,128 @@
+    });
+
+    test('should fail if any mandatory field is missing in form submission', async () => {
+      await act(async () => {
```
I am not a huge fan of `act()` [here](https://medium.com/@AbbasPlusPlus/react-testing-library-understanding-act-and-when-to-use-it-301bd06fd1bc#:~:text=The%20Purpose%20of%20act(),in%20a%20real%20browser%20environment.) you can check a good article.
But this is ok!

dalejo96 commented on client/specs/App.test.tsx:
```diff
@@ -0,0 +1,128 @@
+      mockNavigate.mockClear();
+    });
+
+    test('should fail if any mandatory field is missing in form submission', async () => {
```
avoid usage of should, in test please.

dalejo96 reviewed (changes requested):
Good work!
This is taking form.
Check the comments and let me know if you have any doubts.

BarbDMC committed:
refactor: Improve variables names and descriptions

BarbDMC committed:
refactor: standar axios imports

CristopherDavid commented on client/.gitignore:
```diff
@@ -0,0 +1,23 @@
+
+# misc
+.DS_Store
+.env.local
+.env.development.local
+.env.test.local
+.env.production.local
```
seeing these ignored, maybe we can add a .env.example file to the project so users know what vars they need

CristopherDavid commented on client/package.json:
```diff
@@ -0,0 +1,56 @@
+    ]
+  },
+  "browserslist": {
+    "production": [
+      ">0.2%",
+      "not dead",
+      "not op_mini all"
+    ],
+    "development": [
+      "last 1 chrome version",
+      "last 1 firefox version",
+      "last 1 safari version"
+    ]
```
let's replace this with actual verions or remove the unnecesary boilerplate

CristopherDavid commented on client/package.json:
```diff
@@ -0,0 +1,56 @@
+    "build": "tsc && vite build",
+    "preview": "vite preview"
+  },
+  "eslintConfig": {
+    "extends": [
+      "react-app",
+      "react-app/jest"
+    ]
```
let's add an .eslintrc file to put the configs there

CristopherDavid commented on client/specs/__mocks__/styleMock.ts:
```diff
@@ -0,0 +1 @@
+module.exports = {};
```
Let's get rid of this file if we don't need it or add an end of line here

CristopherDavid commented on client/src/hooks/useLogin.ts:
```diff
@@ -0,0 +1,39 @@
+    setError(null);
+
+    try {
+      const response = await axios.post('http://localhost:3000/v1/login', {
+        email,
+        password
+      });
+
+      localStorage.setItem('token', response.data.token);
+      setLoading(false);
+      
+      return response.data.user;
```
Maybe we can use some schema validation or some mechanism of validation here, to make sure the response is what we want. Check [zod](https://zod.dev/)

CristopherDavid commented on client/src/interfaces/userInterfaces.ts:
```diff
@@ -0,0 +1,13 @@
+
+export interface userSignUpInterface extends userDataInterface {
+  password: string,
+};
```
please check end of line on all files, maybe set Prettier to help you with things like this

CristopherDavid commented on package.json:
```diff
@@ -44,6 +45,7 @@
+    "jest-environment-jsdom": "^29.7.0",
     "prisma": "^5.7.0",
     "supertest": "^6.3.3",
     "ts-jest": "^29.1.1",
```
check dependencies, move things like ts-jest to devDependencies

CristopherDavid commented on package.json:
```diff
@@ -58,6 +60,9 @@
     "preset": "ts-jest",
     "testEnvironment": "node",
-    "testTimeout": 20000
+    "testTimeout": 20000,
+    "moduleNameMapper": {
+      "\\.(css|scss)$": "<rootDir>/client/specs/__mocks__/styleMock.ts"
+    }
```
I don't feel like we need to mock the styles 🤔 wdyt @dalejo96

BarbDMC replied:
I did this because I was having trouble running client tests without mocking styles modules.

CristopherDavid reviewed (changes requested):
Looks good, so far. I left some comments, let me know if you have questions

BarbDMC committed:
build(client): move testing dependencies to devDependencies

the dependencies testing-library jest-dom, react, user-event and vite dependecies were moved to devDependencies

BarbDMC committed:
chore: add .eslint config to client and formating

a .eslint file config was added to client,  styles were fixed  and clients files were correctly formated

BarbDMC committed:
refactor: rename interfaces

interfaces were rename to having a capitalize names

BarbDMC committed:
feat: add 404 page

404 page was added

BarbDMC committed:
refactor: remove unnecessary div container

BarbDMC committed:
feat: handling non axios errors

BarbDMC committed:
feat(SignUp): use reducer and react form

the signup page was change to use reducer and react form hooks to improve data handling

BarbDMC committed:
refactor(SignUp): Fix TS warning

A type warning from typescript was fixed

BarbDMC committed:
chore: vsconfig were added to gitignore file
